lib/logger/index.ts

```diff
@@ -109,69 +133,145 @@ const loggerLevels: bunyan.LogLevelString[] = [
+  }
+
+  addStream(stream: bunyan.Stream): void {
+    this.bunyanLogger.addStream(withSanitizer(stream));
+  }
+
+  childLogger(): RenovateLogger {
+    return new RenovateLogger(
+      this.bunyanLogger.child({}),
+      this.context,
+      this.meta,
+    );
+  }
+
+  trace = this.log.bind(this, 'trace');
+  debug = this.log.bind(this, 'debug');
+  info = this.log.bind(this, 'info');
+  warn = this.log.bind(this, 'warn');
+  error = this.log.bind(this, 'error');
+  fatal = this.log.bind(this, 'fatal');
+
+  once = this.logger.once;
 
-loggerLevels.forEach((loggerLevel) => {
-  logger[loggerLevel] = logFactory(loggerLevel) as never;
+  get logContext(): string {
+    return this.context;
+  }
+
+  set logContext(context: string) {
+    this.context = context;
+  }
+
+  setMeta(obj: Record<string, unknown>): void {
+    this.meta = { ...obj };
+  }
+
+  addMeta(obj: Record<string, unknown>): void {
+    this.meta = { ...this.meta, ...obj };
+  }
 
-  const logOnceFn = (p1: any, p2: any): void => {
-    once(() => {
-      const logFn = logger[loggerLevel];
-      if (is.undefined(p2)) {
-        logFn(p1);
-      } else {
-        logFn(p1, p2);
+  removeMeta(fields: string[]): void {
+    for (const key of Object.keys(this.meta)) {
+      if (fields.includes(key)) {
+        delete this.meta[key];
       }
-    }, logOnceFn);
-  };
-  logger.once[loggerLevel] = logOnceFn as never;
-});
+    }
+  }
 
-const logFile = getEnv('LOG_FILE');
-// istanbul ignore if: not easily testable
-if (is.string(logFile)) {
-  // ensure log file directory exists
-  const directoryName = upath.dirname(logFile);
-  fs.ensureDirSync(directoryName);
+  private logFactory(level: bunyan.LogLevelString): loggerFunction {
+    return logFactory(this.bunyanLogger, level, this.meta, this.context);
+  }
 
-  addStream({
-    name: 'logfile',
-    path: logFile,
-    level: validateLogLevel(getEnv('LOG_FILE_LEVEL'), 'debug'),
-  });
+  private logOnceFn(level: bunyan.LogLevelString): loggerFunction {
+    const logOnceFn = (p1: string | Record<string, any>, p2?: string): void => {
+      once(() => {
+        this.log(level, p1, p2);
+      }, logOnceFn);
+    };
+    return logOnceFn;
+  }
+
+  private log(
+    level: bunyan.LogLevelString,
+    p1: string | Record<string, any>,
+    p2?: string,
+  ): void {
+    const logFn = this.logger[level];
+    if (is.string(p1)) {
+      logFn(p1);
+    } else {
+      logFn(p1, p2);
+    }
+  }
 }
 
+const defaultStreams = createDefaultStreams(
+  stdoutLevel,
+  problems,
+  getEnv('LOG_FILE'),
+);
+const bunyanLogger = bunyan.createLogger({
+  name: 'renovate',
+  serializers: {
+    body: configSerializer,
+    cmd: cmdSerializer,
+    config: configSerializer,
+    migratedConfig: configSerializer,
+    originalConfig: configSerializer,
+    presetConfig: configSerializer,
+    oldConfig: configSerializer,
+    newConfig: configSerializer,
+    err: errSerializer,
+  },
+  streams: defaultStreams.map(withSanitizer),
+});
```

Comment:
can't these be also moved to the logger class? and the constructor more parameterized? that way the logger would be much easier testable (when the class is in a seperate file too.

Reply:
we currently inject the `bunyan` logger instance via the constructor. This approach maximizes flexibility for the user, allowing them to configure the logger as needed. It also keeps the logger class minimal, as it doesnt need to manage details like the `problem` stream or other specifics of the injected logger.


for example, if we move this to the logger class, it would need to expose additional methods, such as `getProblems`, etc.

i prefer to keep this approach, as it is the most generic and flexible for the caller, if thats ok with you :smile:

Reply:
whats more, for the `problem` stream as an example, we could either pass it in the constructor or create it ourselves inside the class.
but what if the caller doesnt need a `problem` stream at all?

i truly believe that leaving all these details outside the logger class is much cleaner while still providing all the flexibility to the user.